modules/siddhi-core/src/main/java/io/siddhi/core/attributes/UserAttributeKey.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.siddhi.core.attributes;
+
+/**
+ * Instead of using plain objects, or an enum, as keys to user data, instances
+ * of {@link UserAttributeKey} carry type information with them.  At run time,
+ * makes no difference (type erasure) but makes the developer experience sightly
+ * better as avoids unnecessary casts when consuming values.  For example:
+ *
+ * <pre>{@code
+ * UserAttribute<MyData> key = new UserAttribute("myDataKey");
+ * SiddhiManager manager = new SiddhiManager();
+ * manager.getUserData().put(key, new MyData());
+ *
+ * ...
+ * // Casting and types are inferred since key is defined with <MyData>
+ * MyData info = manager.getUserData.get(key);
+ * }
+ * </pre>
```

Comment:
Sorry for the delay in the review. This looks good. But here the user should anyway need to know what type of `UserAttribute<MyData> key` they need to create, hence I believe using `UserAttribute`s is unnecessary. Therefore can you please modify this as a simple concurrent Hashmap, and change the map reference as `properties` instead of `userAttributes` (this way we can also use it to pass other parameters, other than user attributes in the future). Please send the PR with the proposed changes we will be able to merge. 
Thanks for your contribution.